File: 54.FloydWarshallAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio        */
/* -----------------------------------------------------------------------  */
/*    Online Compiler (Java): https://onecompiler.com/java                     */
/*    Online Compiler (JS)  : https://onecompiler.com/javascript               */
/*    Code Formatter (TP)   : https://www.tutorialspoint.com/online_java_formatter.htm */
/* -----------------------------------------------------------------------  */

public class FloydWarshallDemo {
    public static void main(String[] args) {

        // Step 1: Define a large number to represent "infinity" for unreachable nodes
        int INF = 99999;

        // Step 2: Define the graph as an adjacency matrix
        int[][] graph = {
            {0, 5, INF, 10},
            {INF, 0, 3, INF},
            {INF, INF, 0, 1},
            {INF, INF, INF, 0}
        };

        // Step 3: Number of vertices in the graph
        int V = graph.length;

        // Step 4: Create a distance matrix (initially a copy of the graph)
        int[][] dist = new int[V][V];
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                dist[i][j] = graph[i][j];
            }
        }

        // Step 5: Apply Floyd Warshall algorithm
        for (int k = 0; k < V; k++) {             // For each intermediate vertex
            for (int i = 0; i < V; i++) {         // For each source vertex
                for (int j = 0; j < V; j++) {     // For each destination vertex
                    // If the path through vertex k is shorter, update dist[i][j]
                    if (dist[i][k] + dist[k][j] < dist[i][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                    }
                }
            }
        }

        // Step 6: Print the final shortest distance matrix
        System.out.println("Shortest distances between every pair of vertices:");
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                if (dist[i][j] == INF)
                    System.out.print("INF ");
                else
                    System.out.print(dist[i][j] + " ");
            }
            System.out.println();
        }
    }
}

File: 54.FloydWarshallAlgo.py
# The Floyd Warshall Algorithm is a dynamic programming algorithm used to find the shortest
# paths between all pairs of vertices in a weighted graph.
# It works with both directed and undirected graphs, and can handle 
# negative edge weights (but not negative cycles).

# Explanation

# The main idea is that the algorithm tries to improve the 
# shortest distance between any two vertices (i, j) by checking whether there exists an 
# intermediate vertex k that makes the path from i → k → j shorter than the currently 
# known path from i → j.

# Formally,
# For each vertex k (as an intermediate node), we update the distance matrix:

# dist[i][j] = min(dist[i][j],dist[i][k]+dist[k][j])

# This runs for every combination of (i, j, k) — resulting in a time complexity of O(V³).

INF = float('inf')

graph = [
    [0, 5, INF, 10],
    [INF, 0, 3, INF],
    [INF, INF, 0, 1],
    [INF, INF, INF, 0]
]

V = len(graph)

dist = [row[:] for row in graph]

for k in range(V):
    for i in range(V):
        for j in range(V):
            if dist[i][k] + dist[k][j] < dist[i][j]:
                dist[i][j] = dist[i][k] + dist[k][j]

print("Shortest distances between every pair of vertices:")
for i in range(V):
    for j in range(V):
        if dist[i][j] == INF:
            print("INF", end = " ")
        else:
            print(dist[i][j], end = " ")
    print()

    


